### PR TITLE
Basic Wayland support

### PR DESCRIPTION
**Pull request by yol**
Pull request for review of #1

### CODE REVIEW

**Comment by yol** [comment]
Note: If you want to build it, you will need waylandpp from https://github.com/pkerling/waylandpp until NilsBrause/waylandpp#4 and NilsBrause/waylandpp#5 are merged upstream.

File: cmake/platform/linux/mir.cmake [hunk 1]
```diff
@@ -1,2 +1,2 @@
-set(PLATFORM_REQUIRED_DEPS OpenGl EGL Mir LibDRM)
+set(PLATFORM_REQUIRED_DEPS OpenGl EGL Mir LibDRM Xkbcommon)
```
Comment (FernetMenta) [inline cmake/platform/linux/mir.cmake]: are you sure you want to touch mir in this early phase?

Comment (yol) [reply cmake/platform/linux/mir.cmake]: Changed the xkbcommon lookup so I don't forget about it - mir and wayland both have to convert xkb keysyms to kodi XBMCKey and I copied the map from Mir. We can save it for later though since it will continue to work as-is.

File: xbmc/input/keyboard/XkbcommonKeymap.cpp [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#if defined(HAVE_XKBCOMMON)
```
Comment (FernetMenta) [inline xbmc/input/keyboard/XkbcommonKeymap.cpp]: that type of guards we want to get rid of with cmake. only include the file on the platform it is required
I would put it into windowing, not input. keep in mind that we want to have headless kodi some day. this input folder is quite a mess whereas windowing is already cleaned up

File: xbmc/utils/EGLUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "EGLUtils.h"
+#include "log.h"
+
+#if defined(HAVE_LIBEGL)
```
Comment (FernetMenta) [inline xbmc/utils/EGLUtils.cpp]: only include file for build with EGL

File: xbmc/input/keyboard/XkbcommonKeymap.cpp (hunk 2 above)
Comment (yol) [reply xbmc/input/keyboard/XkbcommonKeymap.cpp]: The header should not be a problem, but how do I integrate the .cpp file so that it only compiles on those platforms? Should I use if() like in windowing/CMakeLists.txt?

File: xbmc/utils/StringUtils.h [hunk 4]
```diff
@@ -134,9 +134,87 @@ class StringUtils
    \param delimiter Delimiter to be used to split the input string
    \param iMaxStrings (optional) Maximum number of splitted strings
    */
-  static std::vector<std::string> Split(const std::string& input, const std::string& delimiter, unsigned int iMaxStrings = 0);
-  static std::vector<std::string> Split(const std::string& input, const char delimiter, size_t iMaxStrings = 0);
-  static std::vector<std::string> Split(const std::string& input, const std::vector<std::string> &delimiters);
+  static std::vector<std::string> Split(const std::string& input, const std::string& delimiter, unsigned int iMaxStrings = 0)
```
Comment (FernetMenta) [inline xbmc/utils/StringUtils.h]: why do you put bodies into the header?
how is that related to wayland?

File: xbmc/windowing/WinEvents.cpp [hunk 5]
```diff
@@ -39,6 +39,10 @@
 #include "android/WinEventsAndroid.h"
 #define WinEventsType CWinEventsAndroid
 
+#elif defined(TARGET_LINUX) && defined(HAVE_WAYLAND)
```
Comment (FernetMenta) [inline xbmc/windowing/WinEvents.cpp]: TARGET_LINUX seems superfluous

File: xbmc/utils/StringUtils.h (hunk 4 above)
Comment (yol) [reply xbmc/utils/StringUtils.h]: I can put Split() back in the .cpp, but SplitTo() would have to stay because it is a template.
I wanted to use a std::set for the EGL extensions because it is more appropriate than a std::vector. But Split() only returned a std::vector which I would then have to copy to a std::set, which seemed pointless to me.
If you want to avoid the StringUtils change I can revert it and copy the vector to the set after splitting.

File: xbmc/windowing/wayland/GLContextEGL.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+bool CGLContextEGL::CreateDisplay(wayland::display_t& display,
+                                  EGLint renderable_type,
+                                  EGLint rendering_api)
+{
+  EGLint neglconfigs = 0;
+  int major, minor;
+
+  EGLint attribs[] ={
```
Comment (FernetMenta) [inline xbmc/windowing/wayland/GLContextEGL.cpp:63]: no alpha or depth required here?

File: cmake/platform/linux/mir.cmake (hunk 1 above)
Comment (yol) [reply cmake/platform/linux/mir.cmake]: reverted

File: xbmc/input/keyboard/XkbcommonKeymap.cpp (hunk 2 above)
Comment (yol) [reply xbmc/input/keyboard/XkbcommonKeymap.cpp]: done

File: xbmc/utils/EGLUtils.cpp (hunk 3 above)
Comment (yol) [reply xbmc/utils/EGLUtils.cpp]: done

File: xbmc/windowing/WinEvents.cpp (hunk 5 above)
Comment (yol) [reply xbmc/windowing/WinEvents.cpp]: You're right, removed

File: xbmc/windowing/wayland/WinEventsWayland.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  g_WlDisplay = display;
+}
+
+size_t CWinEventsWayland::GetQueueSize()
+{
+  // FIXME Where is this function used?
+  throw std::runtime_error("Not implemented");
+}
+
+// FIXME Why is the return value of this ignored everywhere?
```
Comment (FernetMenta) [inline xbmc/windowing/wayland/WinEventsWayland.cpp]: MessagePump is supposed to die. In some future version rendering should run on its own thread and this thread can send events to application.

File: xbmc/windowing/wayland/GLContextEGL.cpp (hunk 6 above)
Comment (yol) [reply xbmc/windowing/wayland/GLContextEGL.cpp:63]: I admit this is copypaste from Mir. It does seem to work without depth buffer, but we should probably request one anyway. But why do you want alpha?

File: xbmc/windowing/wayland/WinEventsWayland.cpp (hunk 7 above)
Comment (yol) [reply xbmc/windowing/wayland/WinEventsWayland.cpp]: I can put the Wayland message dispatcher on an extra thread and have it only push events to CWinEvents like Mir does. This would make it easier to migrate away from MessagePump when it dies.

File: xbmc/windowing/wayland/WinEventsWayland.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  
+  return true;
+}
+
+void CWinEventsWayland::MessagePush(XBMC_Event* ev)
+{
+  // Currently, this is only called by other Wayland-related code which is in turn
+  // called from g_WlDisplay->dispatch_pending() in MessagePump() on the same thread
+  // so no locking is necessary
+  m_queue.emplace(*ev);
```
Comment (FernetMenta) [inline xbmc/windowing/wayland/WinEventsWayland.cpp]: maybe you can call g_application.OnEvent(event) derectly and drop MessagePump

Comment (yol) [reply xbmc/windowing/wayland/WinEventsWayland.cpp]: Can OnEvent() be called from all threads?

File: xbmc/windowing/wayland/WinEventsWayland.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+{
+namespace WAYLAND
+{
+
+class CWinEventsWayland : public IWinEvents
+{
+public:
+  virtual std::size_t GetQueueSize() override;
+  virtual bool MessagePump() override;
+  virtual void Notify(const Observable& obs, const ObservableMessage msg) override;
```
Comment (FernetMenta) [inline xbmc/windowing/wayland/WinEventsWayland.h]: what is this?
I don't recomment using Kodi's observable. Its has been abused all over the place and you never know what has hooked into it. the call can even block and go out to addons.

Comment (yol) [reply xbmc/windowing/wayland/WinEventsWayland.h]: Unused. Removed. It could probably be dropped from IWinEvents, too, none of the other windowing systems use it.

File: xbmc/windowing/wayland/WinEventsWayland.cpp (hunk 8 above)
Comment (FernetMenta) [reply xbmc/windowing/wayland/WinEventsWayland.cpp]: currently not but this can be easily changed by using CApplcationMessenger. that originally was designed to have events processed by the main thread.

File: xbmc/input/keyboard/XkbcommonKeymap.cpp (hunk 2 above)
Comment (FernetMenta) [reply xbmc/input/keyboard/XkbcommonKeymap.cpp]: I guess you found this, right? https://github.com/xbmc/xbmc/pull/12134

File: xbmc/utils/StringUtils.h (hunk 4 above)
Comment (FernetMenta) [reply xbmc/utils/StringUtils.h]: @notspiff what is your opinion on this?

Comment (notspiff) [reply xbmc/utils/StringUtils.h]: templates in headers is a necessary evil. the other is just evil. having container freedom is a valid reason to use a template for sure.

File: xbmc/windowing/wayland/GLContextEGL.cpp (hunk 6 above)
Comment (FernetMenta) [reply xbmc/windowing/wayland/GLContextEGL.cpp:63]: I only compared with X11. Not sure if we want alpha :) maybe this is a relict from the times the visual was required to be compatible with SDL.

File: xbmc/windowing/wayland/WinEventsWayland.cpp (hunk 7 above)
Comment (FernetMenta) [reply xbmc/windowing/wayland/WinEventsWayland.cpp]: However you like. I just wanted to make sure that you are aware of this.

File: xbmc/windowing/wayland/WinEventsWayland.h (hunk 9 above)
Comment (FernetMenta) [reply xbmc/windowing/wayland/WinEventsWayland.h]: yes, it's only confusing

File: xbmc/input/keyboard/XkbcommonKeymap.cpp (hunk 2 above)
Comment (yol) [reply xbmc/input/keyboard/XkbcommonKeymap.cpp]: Yes, using CORE_PLATFORM_NAME for conditional include now. My only gripe is that if users set CORE_PLATFORM_NAME to e.g. `Wayland` (with a capital `W`), it won't compile since CMake string comparisons are case-sensitive. But maybe I should submit a PR for normalizing the variable in the main Kodi repo.

File: xbmc/windowing/wayland/GLContextEGL.cpp (hunk 6 above)
Comment (yol) [reply xbmc/windowing/wayland/GLContextEGL.cpp:63]: Depth buffer added - music visualization was indeed looking funny without it :-)
My understanding is that we would need alpha buffer on the surface only if the Kodi window should have transparency, which I think is not the case

File: xbmc/utils/StringUtils.h (hunk 4 above)
Comment (yol) [reply xbmc/utils/StringUtils.h]: Split() moved

**Comment by yol** [comment]
Just noticed that this does not build with gcc. Will fix.

**Comment by FernetMenta** [comment]
good work so far. are you feeling comfortable now with wayland-client?

**Comment by yol** [comment]
I'm not using libwayland-client. Did you mean waylandpp?

**Comment by FernetMenta** [comment]
ahh yes. I noticed there is activity on the repo. great!

**Comment by yol** [comment]
Yes, the author has generally been responsive and merged the PRs I put out in a short time, so using waylandpp should be fine for the time being.

**Comment by yol** [comment]
Rebased on master, I'll open a new PR to get rid of the extra commits.
